### DISCUSSION TITLE
Do Multi-Frame Files benefit from Interframe Compression?

### DISCUSSION THREAD

**Discussion by william-silversmith** [libjxl/libjxl#3634] [Q&A]
Hi! Thanks for all the great work on this reference implementation!

I'm working with stacks of electron microscope images that can looks pretty similar from frame to frame. We are looking into using JPEG-XL for archiving very large scientific datasets (potentially petapixels). I was experimenting with it and noticed that JPEG-XL supports frames, including compositing frames.

However, after going through the documentation, it wasn't clear to me if there was an obvious way to benefit from inter-frame compression (though there was a suggestion that frames could be [difference coded](https://ds.jpeg.org/whitepapers/jpeg-xl-whitepaper.pdf) without motion compensation). I suspect the overall compression factor will increase substantially if this 3D data could be exploited.

I did some experiments using the [`imagecodecs`](https://github.com/cgohlke/imagecodecs) python interface to the reference implementation. I did not observe a significant benefit from organizing the file as a long 2D image (with each frame placed next to the previous one on the Y axis) versus using the multi-frame capability.

**Comment by lfz** [comment]
same findings here. even the same picture, repeat N times, the size grows linearly

**Reply by lfz** [reply]
what is your best practice now @william-silversmith

**Reply by william-silversmith** [reply]
I ended up using the same compression trick as before, take a 3D image and turn it into a 2D image with different slices offset in X or Y. I did some research and there is an obscure lossless comression codec or two out there that considers interframe coding, but the vast majority of them are lossy video compressors. I ran into some decoding issue with multiframe too that resolved with 2D as well, but I can't remember what it was now.

One reason lossless compressors are less likely to do interframe coding is small errors can propagate further, which hinders their utility for archival purposes. That is something I gathered from ffmpeg lossless compression codec [ffv1's](https://github.com/FFmpeg/FFV1/blob/master/ffv1.md) documentation.

**Comment by lfz** [comment]
@william-silversmith thank you for your sharing!

I also did the trial for your trick before with jpeg, which I found useless, I conducted another experiment with jpeg xl just now, the result also shows no benefit?

my image is CT, 512 x 512, uint16 gray scale, I choose 9 consecutive slices and organize them into 3x3 mosaic, compared with 9 independent compression, the result of mosaic is not better than independent(except for decoding time)


| Method                         |   Original Size (MB) |   Total Compressed Size (MB) |   Compression Time (s) |   Decompression Time (s) |   Compression Ratio | Is Valid   |
|:-------------------------------|---------------------:|-----------------------------:|-----------------------:|-------------------------:|--------------------:|:-----------|
| Mosaic 3x3                     |                  4.5 |                      1.0685  |                  2.239 |                    0.069 |                4.21 | True       |
| Single Images (9 compressions) |                  4.5 |                      1.07788 |                  2.065 |                    0.114 |                4.17 | True       |

**Comment by Galaxy4594** [comment]
Yes, I've seen this happen when compressing animated image formats (especially ones with static backgrounds) to JXL, file size does decrease quite a bit. The same logic should apply to multi-frame JXLs as well. The current problem is that libjxl doesn't really do any kind of intra-frame compression at the moment, so to get around this issue I use apngasm to do the intra-frame compression for me, and then losslessly convert the APNG to JXL.

I do hope the devs add some kind of a dedicated tool for making animated/mult-frame JXLs with proper intra-frame compression. But that seems to be a goal in the far future.

**Comment by Kvach-Lukich** [comment]
If the most efficient method ends up being to store multiple frames on a single canvas and the frames are similar, would it make sense to first compute the difference between frames, store the first frame, and then store all subsequent frames as the difference plus a bitmask indicating where negative values occur (which is a large amount of data — one bit per pixel on each channel), but with the benefit that the difference itself would look like a very smooth gradient?

Or is the codec smart enough to optimize the entire large canvas and detect identical fragments even if they are far apart from each other?
